### PR TITLE
Support iterables in adapt() for preprocessing layers

### LINKED ISSUE #22156: Support arbitrary Python iterables in preprocessing layer adapt() methods
Currently, most preprocessing layers (`FeatureSpace`, `IndexLookup`, `TextVectorization`, `Discretization`) only accept `tf.data.Dataset` objects in their `adapt()` methods. This creates friction for users working with JAX or PyTorch backends who typically have data in other formats like lists, generators, or custom iterables.

### Current Limitation

```python
# This works
import tensorflow as tf
dataset = tf.data.Dataset.from_tensor_slices(data)
layer.adapt(dataset)

# This doesn't work
layer.adapt(["apple", "banana", "cherry"])  # Error
```

## Proposed Solution

Enable preprocessing layers to accept arbitrary Python iterables (lists, generators, custom iterables) in their `adapt()` methods, similar to how `Normalization` already works.

### Expected Behavior

```python
# Lists should work
layer.adapt(["apple", "banana", "cherry"])  

# Generators should work (memory efficient!)
def data_generator():
    for item in huge_dataset:
        yield item
layer.adapt(data_generator())  

# Custom iterables should work
layer.adapt(custom_iterable_object)  
```

## Benefits

1. **Backend Agnostic**: JAX and PyTorch users can use native Python data structures without converting to TensorFlow datasets
2. **Memory Efficient**: Generators allow processing large datasets without loading everything into memory
3. **Consistent API**: All preprocessing layers would behave consistently with `Normalization`
4. **Backward Compatible**: Existing code using `tf.data.Dataset` continues to work
5. **Pythonic**: More natural for users coming from standard Python/NumPy workflows

## Affected Layers

- `FeatureSpace`
- `StringLookup` / `IntegerLookup` (via `IndexLookup`)
- `TextVectorization`
- `Discretization`

### 1. Processing large text files with generators
```python
def text_generator():
    with open("large_corpus.txt") as f:
        for line in f:
            yield line.strip()

vectorizer = TextVectorization()
vectorizer.adapt(text_generator(), batch_size=64)
```

### 2. JAX/PyTorch users with native data structures
```python
# JAX users with lists
data = ["text1", "text2", "text3"]
layer.adapt(data)

# PyTorch users with custom iterables
layer.adapt(pytorch_dataloader)
```

### 3. Quick prototyping and testing
```python
# Simple list for quick testing
lookup = StringLookup()
lookup.adapt(["cat", "dog", "bird"])
```

## Related

This aligns with Keras 3's goal of being backend-agnostic and improving multi-backend compatibility. The idea for this came as talked about in /keras-team/keras/issues/18442.

### PR DESCRIPTION
**Pull request by rstar327**
## Description

Enable `Discretization`, `IndexLookup` (base for `StringLookup`/`IntegerLookup`), and `TextVectorization` preprocessing layers to accept arbitrary Python iterables (generators, custom iterables) in their `adapt()` methods, in addition to the already-supported `tf.data.Dataset` and NumPy arrays.

This aligns these layers with `Normalization`, which already supports iterables, and improves backend-agnostic usage for JAX and PyTorch users who may not have data in `tf.data.Dataset` format.

Fixes #22156

## Changes

- **`Discretization.adapt()`**: Added iterable branch that iterates batches and calls `update_state()` on each.
- **`IndexLookup.adapt()`**: Added iterable branch; lists/tuples/arrays still go through the existing `ensure_tensor` path for backward compatibility.
- **`TextVectorization.adapt()`**: Same pattern as `IndexLookup`.
- Added tests for generator and list inputs across `Discretization`, `TextVectorization`, and `StringLookup`.

## Example

```python
# Generator support (memory-efficient for large datasets)
def data_gen():
    for chunk in load_large_dataset():
        yield chunk

layer = keras.layers.Discretization(num_bins=10)
layer.adapt(data_gen())

# Works for text layers too
def text_gen():
    with open("corpus.txt") as f:
        while batch := [f.readline() for _ in range(64)]:
            yield batch

vectorizer = keras.layers.TextVectorization()
vectorizer.adapt(text_gen())
```

### CODE REVIEW

File: keras/src/layers/preprocessing/discretization.py [hunk 1]
```diff
@@ -171,8 +171,9 @@ def adapt(self, data, steps=None):
 
         Arguments:
             data: The data to train on. It can be passed either as a
-                batched `tf.data.Dataset`,
-                or as a NumPy array.
+                batched `tf.data.Dataset`, as a NumPy array, or as any
+                iterable of batches (e.g. a list of arrays or a generator
+                yielding batches).
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization.py]: Mention "Grain dataset" too.

File: keras/src/layers/preprocessing/discretization.py [hunk 2]
```diff
@@ -192,6 +193,11 @@ def adapt(self, data, steps=None):
                 data = data.take(steps)
             for batch in data:
                 self.update_state(batch)
+        elif isinstance(data, np.ndarray) or backend.is_tensor(data):
+            self.update_state(data)
+        elif hasattr(data, "__iter__"):
+            for batch in data:
+                self.update_state(batch)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization.py]: We should honor the `steps` argument here.

File: keras/src/layers/preprocessing/discretization.py [hunk 3]
```diff
@@ -192,6 +193,11 @@ def adapt(self, data, steps=None):
                 data = data.take(steps)
             for batch in data:
                 self.update_state(batch)
+        elif isinstance(data, np.ndarray) or backend.is_tensor(data):
+            self.update_state(data)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization.py]: I think we should add `or tf.is_tensor(data)` to the `if` here to keep the same behavior as before.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 4]
```diff
@@ -474,3 +474,29 @@ def test_compute_output_shape_num_bins(
...
+        ]
+        layer = layers.Discretization(num_bins=3)
+        layer.adapt(batches)
+        self.assertLen(layer.bin_boundaries, 2)
+
+    def test_adapt_with_numpy_array(self):
+        data = np.random.uniform(0, 10, size=(50,))
+        layer = layers.Discretization(num_bins=4)
+        layer.adapt(data)
+        self.assertLen(layer.bin_boundaries, 3)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: This is already covered by the test line 33. So remove this test. But you can move the `self.assertLen(layer.bin_boundaries, 3)` to the existing test line 38.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 5]
```diff
@@ -474,3 +474,29 @@ def test_compute_output_shape_num_bins(
...
             f"Failed for num_bins={num_bins}, mode={output_mode}",
         )
+
+    def test_adapt_with_generator(self):
+        def data_gen():
+            for _ in range(5):
+                yield np.random.uniform(0, 10, size=(20,))
+
+        layer = layers.Discretization(num_bins=4)
+        layer.adapt(data_gen())
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Also add a test with an infinite iterator and a `steps` argument:

```python
        def data_gen():
            while True:
                yield np.random.uniform(0, 10, size=(20,))
```

File: keras/src/layers/preprocessing/discretization_test.py [hunk 6]
```diff
@@ -474,3 +474,29 @@ def test_compute_output_shape_num_bins(
...
+        self.assertEqual(output.shape, (1, 3))
+
+    def test_adapt_with_list_of_arrays(self):
+        batches = [
+            np.array([1.0, 2.0, 3.0]),
+            np.array([4.0, 5.0, 6.0]),
+        ]
+        layer = layers.Discretization(num_bins=3)
+        layer.adapt(batches)
+        self.assertLen(layer.bin_boundaries, 2)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Also add a test with a Grain dataset.

File: keras/src/layers/preprocessing/index_lookup.py [hunk 7]
```diff
@@ -617,13 +617,21 @@ def adapt(self, data, steps=None):
...
+        elif isinstance(data, (list, tuple, np.ndarray)):
             data = tf_utils.ensure_tensor(data, dtype=self.vocabulary_dtype)
             if data.shape.rank == 1:
                 # A plain list of strings
                 # is treated as as many documents
                 data = tf.expand_dims(data, -1)
             self.update_state(data)
+        elif hasattr(data, "__iter__"):
+            for batch in data:
+                self.update_state(batch)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/index_lookup.py]: We should honor the `step` argument here.

File: keras/src/layers/preprocessing/index_lookup.py [hunk 8]
```diff
@@ -617,13 +617,21 @@ def adapt(self, data, steps=None):
...
+        else:
+            data = tf_utils.ensure_tensor(data, dtype=self.vocabulary_dtype)
+            if data.shape.rank == 1:
+                data = tf.expand_dims(data, -1)
+            self.update_state(data)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/index_lookup.py]: This will have a different behavior than before for TensorFlow tensors, that's because they're iterable.

So the check line 620 should add `or backend.is_tensor(data) or tf.is_tensor(data)`.

There is also an opportunity to remove the redundant code between the 2 blocks.

File: keras/src/layers/preprocessing/string_lookup_test.py [hunk 9]
```diff
@@ -261,3 +261,21 @@ def test_one_hot_symbolic_output_shape_nested_input(self):
...
+        vocab = layer.get_vocabulary()
+        self.assertIn("cat", [str(v) for v in vocab])
+        self.assertIn("dog", [str(v) for v in vocab])
+        self.assertIn("bird", [str(v) for v in vocab])
+
+    def test_adapt_with_list_backward_compat(self):
+        layer = layers.StringLookup()
+        layer.adapt(["cat", "dog", "bird"])
+        vocab = layer.get_vocabulary()
+        self.assertIn("cat", [str(v) for v in vocab])
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/string_lookup_test.py]: This is already covered by the test line 61.

But feel free to move the assert line 68.

File: keras/src/layers/preprocessing/string_lookup_test.py [hunk 10]
```diff
@@ -261,3 +261,21 @@ def test_one_hot_symbolic_output_shape_nested_input(self):
             tuple(symbolic_output.shape)[1:],
             eager_output.shape[1:],
         )
+
+    def test_adapt_with_generator(self):
+        def vocab_gen():
+            yield ["cat", "dog"]
+            yield ["bird", "cat"]
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/string_lookup_test.py]: Also add:
- a test with an infinite generator and a `steps` argument.
- a test with a Grain dataset

File: keras/src/layers/preprocessing/text_vectorization.py [hunk 11]
```diff
@@ -419,13 +420,21 @@ def adapt(self, data, batch_size=None, steps=None):
...
+        elif isinstance(data, (list, tuple, np.ndarray)):
             data = tf_utils.ensure_tensor(data, dtype="string")
             if data.shape.rank == 1:
                 # A plain list of strings
                 # is treated as as many documents
                 data = tf.expand_dims(data, -1)
             self.update_state(data)
+        elif hasattr(data, "__iter__"):
+            for batch in data:
+                self.update_state(batch)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization.py]: We should honor the `step` argument here.

File: keras/src/layers/preprocessing/text_vectorization.py [hunk 12]
```diff
@@ -419,13 +420,21 @@ def adapt(self, data, batch_size=None, steps=None):
...
+        else:
+            data = tf_utils.ensure_tensor(data, dtype="string")
+            if data.shape.rank == 1:
+                data = tf.expand_dims(data, -1)
+            self.update_state(data)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization.py]: This will have a different behavior than before for TensorFlow tensors, that's because they're iterable.

So the check line 620 should add `or backend.is_tensor(data) or tf.is_tensor(data)`.

There is also an opportunity to remove the redundant code between the 2 blocks.

File: keras/src/layers/preprocessing/text_vectorization_test.py [hunk 13]
```diff
@@ -524,3 +524,20 @@ def test_output_sequence_length_with_non_int_mode(self):
                 output_mode="multi_hot",
                 output_sequence_length=5,
             )
+
+    def test_adapt_with_generator(self):
+        def text_gen():
+            yield ["hello world", "foo bar"]
+            yield ["baz qux", "hello foo"]
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization_test.py]: Also add:
- a test with an infinite generator and a `steps` argument.
- a test with a Grain dataset

File: keras/src/layers/preprocessing/text_vectorization_test.py [hunk 14]
```diff
@@ -524,3 +524,20 @@ def test_output_sequence_length_with_non_int_mode(self):
...
+        layer.adapt(text_gen())
+        vocab = layer.get_vocabulary()
+        self.assertIn("hello", [str(v) for v in vocab])
+        self.assertIn("foo", [str(v) for v in vocab])
+
+    def test_adapt_with_list_backward_compat(self):
+        layer = layers.TextVectorization()
+        layer.adapt(["hello world", "foo bar", "baz qux"])
+        vocab = layer.get_vocabulary()
+        self.assertIn("hello", [str(v) for v in vocab])
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization_test.py]: This is already covered by the test line 28. So remove this test. But feel free to move the assert to line 37.

File: keras/src/layers/preprocessing/text_vectorization_test.py [hunk 15]
```diff
@@ -524,3 +524,20 @@ def test_output_sequence_length_with_non_int_mode(self):
                 output_mode="multi_hot",
                 output_sequence_length=5,
             )
+
+    def test_adapt_with_generator(self):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization_test.py]: Please move all the tests right after line 41 so that the `adapt` tests are together.

File: keras/src/layers/preprocessing/string_lookup_test.py [hunk 16]
```diff
@@ -261,3 +261,21 @@ def test_one_hot_symbolic_output_shape_nested_input(self):
             tuple(symbolic_output.shape)[1:],
             eager_output.shape[1:],
         )
+
+    def test_adapt_with_generator(self):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/string_lookup_test.py]: Please move all the tests right after line 70 so that the `adapt` tests are together.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 17]
```diff
@@ -474,3 +474,29 @@ def test_compute_output_shape_num_bins(
             expected_shape,
             f"Failed for num_bins={num_bins}, mode={output_mode}",
         )
+
+    def test_adapt_with_generator(self):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Please move all the tests right after line 40 so that the `adapt` tests are together.

**Comment by rstar327** [comment]
Done

File: keras/src/layers/preprocessing/discretization_test.py [hunk 18]
```diff
@@ -37,6 +37,56 @@ def test_adapt_flow(self):
...
+        ]
+        layer = layers.Discretization(num_bins=3)
+        layer.adapt(batches)
+        self.assertLen(layer.bin_boundaries, 2)
+
+    def test_adapt_with_grain_dataset(self):
+        try:
+            import grain
+        except ImportError:
+            self.skipTest("Grain is not installed.")
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Move the `import grain` to the top and remove the `try / except`.

While you cannot assume that `grain` is installed the actual keras code, you can assume it's installed in the unit tests.

File: keras/src/layers/preprocessing/string_lookup_test.py [hunk 19]
```diff
@@ -67,6 +67,50 @@ def test_adapt_flow(self):
...
+        layer = layers.StringLookup()
+        layer.adapt(vocab_gen(), steps=3)
+        vocab = layer.get_vocabulary()
+        self.assertIn("cat", [str(v) for v in vocab])
+
+    def test_adapt_with_grain_dataset(self):
+        try:
+            import grain
+        except ImportError:
+            self.skipTest("Grain is not installed.")
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/string_lookup_test.py]: Move the `import grain` to the top and remove the `try / except`.

While you cannot assume that `grain` is installed the actual keras code, you can assume it's installed in the unit tests.

File: keras/src/layers/preprocessing/text_vectorization_test.py [hunk 20]
```diff
@@ -38,6 +38,49 @@ def test_adapt_flow(self):
...
+        layer = layers.TextVectorization()
+        layer.adapt(text_gen(), steps=3)
+        vocab = layer.get_vocabulary()
+        self.assertIn("hello", [str(v) for v in vocab])
+
+    def test_adapt_with_grain_dataset(self):
+        try:
+            import grain
+        except ImportError:
+            self.skipTest("Grain is not installed.")
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/text_vectorization_test.py]: Move the `import grain` to the top and remove the `try / except`.

While you cannot assume that `grain` is installed the actual keras code, you can assume it's installed in the unit tests.

**Comment by rstar327** [comment]
Done
